rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+ 
+### Persistent Bot Instance Identity
+
+Today, there is no persistent identifier for an individual instance of a Bot.
+Instead, all `tbot` instances are effectively identified solely by their Bot
+identity. There is no easy way to distinguish them. On each renewal, `tbot`
+regenerates the private-public key pair that is used within its certificate and
+there is no other form of unique ID that is persisted across renewals.
+
+This poses a few challenges:
+
+- For the purposes of auditing, it is not possible to trace actions to a
+  specific instance of a Bot.
+- For improving the `token` join method to support multiple Bot instances
+  associated with a single Bot, there is no identifier to correlate with the 
+  generation counter.
+- For analytics purposes, it's difficult for us to track the number of
+  individual Bot instances in use. We cannot easily determine if it's a single
+  very active Bot instance, or many less active Bot instances.
+
+To rectify this, a unique identifier should be established for an instance of
+a Bot.
+
+#### Public Key Fingerprint
+
+One option is to modify the behaviour of `tbot` to persist and reuse the
+keypair across renewals. We could then use a fingerprint of the public key as a
+unique identifier of the Bot instance.
+
+This feels like a natural identifier. It avoids introducing a new attribute to
+certificates as the public key is already encoded within certificates. The
+nature of public-key cryptography also means that this provides the Bot instance
+a way to identify itself without needing an issued certificate.
+
+To mitigate the risk of pre-image attacks, SHA256 will be used to determine the
+fingerprint of the public key. In addition, the full public key should be
+recorded and verified against when authenticating a Bot instance action.
+
+It should be noted that with this technique, rotating the keypair of a `tbot`
+instance would reset the identity of that instance. Rotation of this keypair
+would be unusual and this side effect seems expected.
+
+##### Alternative: UUID Certificate Attribute
+
+On the initial join of a Bot instance, we could generate a UUID to identify that 
+Bot instance and encode this within the certificate. Upon renewals, the UUID
+would be copied from the current certificate and into the new one.
+
+Whilst this is fairly easy to implement for the `token` join method, one
+challenge for the delegated join methods is that rather than renewing, the
+`tbot` instance merely re-joins. As the join RPCs are unauthenticated, the 
+previous certificate of the Bot instance is not readily available. We can
+either:
+
+- Accept this limitation and treat each renewal of a delegated Bot instance as a
+  new Bot instance. This is likely unacceptable and would limit any advantages
+  of this work to the `token` join method.
+- Add support for calling the join RPCs with a client certificate.
+
+This technique could reuse a recently proposed LoginID attribute. This would 
```

Comment:
I'm curious about whether we'll move forward with the proposed idea of LoginIDs from the hackathon - we could even straight up re-use this for join/renews of Machine ID.

Reply:
It looks like there's not really been much movement on this since January. We can probably safely assume we do not want to depend on that.

Reply:
Since it looks like we're dropping this, I'll remove the reference.